### PR TITLE
Elemental

### PR DESCRIPTION
**Pull request by bvanessen**
Added a package for the Elemental distributed linear algebra library.

### CODE REVIEW

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    url      = "https://github.com/elemental/Elemental/archive/v0.87.6.tar.gz"
+
+    version('0.87.6', '9fd29783d45b0a0e27c0df85f548abe9')
+
+    variant('debug', default=False, 
+            description='Builds a debug version of the libraries')
+    variant('shared', default=True, 
+            description='Enables the build of shared libraries')
+    variant('hybrid', default=True, 
+            description='Elemental: make use of OpenMP within MPI packing/unpacking')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: you don't need `Elemental: ` in variant descriptions.

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    variant('c_interface', default=False, 
+            description='Elemental: build C interface')
+    variant('python_package', default=False, 
+            description='Elemental: install Python interface')
+    variant('disable_parmetis', default=True,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: I think we normally keep it positive, I would call it `parmetis` with  description `Enable parMETIS`?

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            description='Elemental: disable quad precision')
+    variant('int64', default=False, 
+            description='Elemental: use 64bit integers')
+    variant('int64_blas', default=False, 
+            description='Elemental: use 64bit integers for BLAS')
+    variant('scalapack', default=False,
+            description='Elemental: build with ScaLAPACK library')
+
+    depends_on('cmake', type='build')
+    depends_on('openblas +openmp', when='~int64_blas')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: you should keep it as  `depends_on('blas')` as it will most likely build with any other `blas`. Unfortunately, until https://github.com/LLNL/spack/issues/1712 is fixed, you can't pass variants to virtual dependencies.

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    depends_on('metis')
+    depends_on('metis +int64', when='+int64')
+    depends_on('mpi')
+    depends_on('netlib-scalapack', when='+scalapack ~int64_blas')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: same here, there are other providers for `scalapack`.

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                         '-DCUSTOM_BLAS_SUFFIX:BOOL=TRUE']),
+            if '+scalapack' in self.spec:
+                args.extend(['-DEL_LAPACK_SUFFIX:STRING={0}'.format((
+                    '_64_' if '+int64_blas' in self.spec else '_')),
+                             '-DCUSTOM_LAPACK_SUFFIX:BOOL=TRUE']),
+        else:
+            if '+scalapack' in self.spec:
+                args.extend(['-DMATH_PATHS:STRING=-L{0}/lib -L{1}/lib'.format(
+                    self.spec['openblas'].prefix, self.spec['netlib-scalapack'].prefix),
+                             '-DMATH_LIBS:STRING=-lopenblas -lscalapack'])
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: you get `blas` object from `blas = spec['blas'].blas_libs` and link flags can be get from `blas.ld_flags`, see other packages which use `blas/lapack`.

Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: same for `scalapack`, `spec['scalapack_libs'].scalapack_libs` will give `LibraryList` object.

Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: what you most likely want here is 
```
        math_libs = spec['scalapack_libs']. scalapack_libs + spec['lapack'].lapack_libs + spec['blas'].blas_libs
```
and then do 
```
math_libs.ld_flags
```

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    homepage = "http://libelemental.org"
+    url      = "https://github.com/elemental/Elemental/archive/v0.87.6.tar.gz"
+
+    version('0.87.6', '9fd29783d45b0a0e27c0df85f548abe9')
+
+    variant('debug', default=False, 
+            description='Builds a debug version of the libraries')
+    variant('shared', default=True, 
+            description='Enables the build of shared libraries')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: With `+shared` enabled, does it also build a static library, or just the shared library?
Is there a way to have it build the static library with position-independent code (PIC)?

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 4 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: Can you explain the reason for `~int64_blas` here?  A little explanatory text behind the logic (in comments) would be helpful.

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 2 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: 1. Normally, we would make the `c_interface` default to True and `python_package` default to False.  The idea is that variants that add dependencies should default to False.

2. Please rename `python_package` to `python`, since this is standard.  And consider renaming `c_interface` to `c`.

Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: If this package builds a Python interface, then you need:
```
depends_on('python')
extends('python')
```
What version(s) of Python does your interface work with?  Is it Python2-only, or Python3-only, or both?  This needs to be specified in the `depends_on('python')'` line (grep for examples...)

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 5 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: Can you clean these up visually a little?  I'd like to see something like:
```
                '-DINSTALL_PYTHON_PACKAGE:BOOL=%s' %
                     ('ON' if '+python_package' in spec else 'OFF')),
```
(Add `spec=self.spec` as the first line of the method)

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    variant('hybrid', default=True, 
+            description='Elemental: make use of OpenMP within MPI packing/unpacking')
+    variant('c_interface', default=False, 
+            description='Elemental: build C interface')
+    variant('python_package', default=False, 
+            description='Elemental: install Python interface')
+    variant('disable_parmetis', default=True, 
+            description='Elemental: disable ParMETIS')
+    variant('disable_quad', default=True, 
+            description='Elemental: disable quad precision')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: I agree on the positive logic.  `diable_parmetis default=True` should be changed to `parmetis default=False`.  Same thing for `disable_quad` turning into `quad` variant.  And the logic should be reversed below, as appropriate.

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    variant('int64', default=False, 
+            description='Elemental: use 64bit integers')
+    variant('int64_blas', default=False, 
+            description='Elemental: use 64bit integers for BLAS')
+    variant('scalapack', default=False,
+            description='Elemental: build with ScaLAPACK library')
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: What would be the cost if these were defaulted to `True`?

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            return 'Debug'
+        else:
+            return 'Release'
+
+    def cmake_args(self):
+        args = ['-DCMAKE_INSTALL_MESSAGE:STRING=LAZY',
+                '-DEL_PREFER_OPENBLAS:BOOL=TRUE',
+                '-DEL_DISABLE_SCALAPACK:BOOL={0}'.format((
+                    'ON' if '+scalapack' in self.spec else 'OFF')),
+                '-DGFORTRAN_LIB=libgfortran.so',
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: Why is this needed?  Normally, you do not have to pass the location or name of `libgfortran.so` to builds.  This looks like a wonky thing in this build; but it will ultimately bite us when someone tries to build with non-GCC.  I'd like to get more explanation of why it is here, then maybe we can find a better way.

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 1 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: Do we have a standard variant name for what is called `+hybrid` here?

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    depends_on('openblas +openmp', when='~int64_blas')
+    depends_on('metis')
+    depends_on('metis +int64', when='+int64')
+    depends_on('mpi')
+    depends_on('netlib-scalapack', when='+scalapack ~int64_blas')
+
+    def build_type(self):
+        """Returns the correct value for the ``CMAKE_BUILD_TYPE`` variable
+        :return: value for ``CMAKE_BUILD_TYPE``
+        """
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: @adamjstewart 
Is this a part of `CMakePackage` now?  Has #2692 gone somewhere?  Ideally, any CMake-standard `build_type` stuff should happen in `CMakePackage`, not in `package.py`.  Can we delete this method?

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                             '-DCUSTOM_LAPACK_SUFFIX:BOOL=TRUE']),
+        else:
+            if '+scalapack' in self.spec:
+                args.extend(['-DMATH_PATHS:STRING=-L{0}/lib -L{1}/lib'.format(
+                    self.spec['openblas'].prefix, self.spec['netlib-scalapack'].prefix),
+                             '-DMATH_LIBS:STRING=-lopenblas -lscalapack'])
+            else:
+                args.extend(['-DMATH_PATHS:STRING=-L{0}/lib'.format(
+                    self.spec['openblas'].prefix),
+                             '-DMATH_LIBS:STRING=-lopenblas'])
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/elemental/package.py]: Again, this is a potential problem.  Different BLAS implemenations give different names to their blas libraries.  This package should really be set up so it can work with any BLAS.

**Comment by citibeth** [comment]
Thank you for submitting a new package, it is truly appreciated!  Don't be intimidated by the vetting process... ;-)

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 10 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/elemental/package.py]: http://spack.readthedocs.io/en/latest/spack.build_systems.html#spack.build_systems.cmake.CMakePackage.build_type

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 1 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: > Do we have a standard variant name for what is called +hybrid here?

since it depends on `mpi` unconditionally, i think it's appropriate to call this variant `openmp` instead of `hybrid`.

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 3 above)
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: since you also use scalapack, then most likely you actually need
```
depends_on('blas)
depends_on('lapack')
```

File: var/spack/repos/builtin/packages/elemental/package.py (hunk 2 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/elemental/package.py]: `extends('python')` _implies_ `depends_on('python')`. You don't need both unless you need to specify a certain version of Python that is required.

**Comment by davydden** [comment]
@bvanessen have a look at `cp2k` or `abinit` which use `scalapack`, `lapack` and `blas`.

**Comment by bvanessen** [comment]
I don’t believe that is correct.  The hybrid variant is used to enable OpenMP threading inside of Elemental with respect to interfacing with MPI.  The openmp variant is related to the threading model in the BLAS library.  I have clarified this in the varients names.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        else:
+            lapack = self.spec['lapack'].lapack_libs
+            blas = self.spec['blas'].blas_libs
+
+            if '+scalapack' in self.spec:
+                scalapack = self.spec['scalapack'].scalapack_libs
+                args.extend(['-DMATH_LIBS:STRING={0} {1}'.format(
+                    (lapack + blas).search_flags, scalapack.search_flags),
+                             '-DMATH_LIBS:STRING={0} {1}'.format(
+                    (lapack + blas).ld_flags.split()[1], scalapack.ld_flags.split()[1])])
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: Are you getting `-lfoo -lbar -lbaz` here? If so, there is `link_flags()`. Note that `ld_flags` may have more than one directory set with `-L`.
(same below)

Also, should the order be the opposite (scalapack first)?

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+        if '+int64_blas' in self.spec:
+            args.extend(['-DEL_BLAS_SUFFIX:STRING={0}'.format((
+                '_64_' if '+int64_blas' in self.spec else '_')),
+                         '-DCUSTOM_BLAS_SUFFIX:BOOL=TRUE']),
+            if '+scalapack' in self.spec:
+                args.extend(['-DEL_LAPACK_SUFFIX:STRING={0}'.format((
+                    '_64_' if '+int64_blas' in self.spec else '_')),
+                             '-DCUSTOM_LAPACK_SUFFIX:BOOL=TRUE']),
+        else:
+            lapack = self.spec['lapack'].lapack_libs
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: you never use them alone, why not haveing `lapack_blas =  self.spec['lapack'].lapack_libs+self.spec['lapack'].blas_libs`?

File: var/spack/repos/builtin/packages/elemental/package.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+                '_64_' if '+int64_blas' in self.spec else '_')),
+                         '-DCUSTOM_BLAS_SUFFIX:BOOL=TRUE']),
+            if '+scalapack' in self.spec:
+                args.extend(['-DEL_LAPACK_SUFFIX:STRING={0}'.format((
+                    '_64_' if '+int64_blas' in self.spec else '_')),
+                             '-DCUSTOM_LAPACK_SUFFIX:BOOL=TRUE']),
+        else:
+            lapack_blas = self.spec['lapack'].lapack_libs + self.spec['blas'].blas_libs
+
+            if '+scalapack' in self.spec:
```
Comment (davydden) [inline var/spack/repos/builtin/packages/elemental/package.py]: you can actually make it even more concise by
```
math_libs = self.spec['lapack'].lapack_libs + self.spec['blas'].blas_libs
if '+scalapack' in self.spec:
    math_libs = self.spec['scalapack'].scalapack_libs + math_libs
```
and then use `math_libs` to get `search_flags` and `link_flags`

**Comment by bvanessen** [comment]
Thanks for the feedback.  How do I set the PYTHON_SITE_PACKAGES variable for elemental.  Basically, I need to know where spack is installing python packages so that I can give that to the build environment.  I assume that this info is available through some spack environment variable, but I don’t know which one.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by davydden** [comment]
> How do I set the PYTHON_SITE_PACKAGES variable for elemental

i don't use python, hopefully others can comment.

**Comment by adamjstewart** [comment]
Try `site_packages_dir`. See `libxml2` for an example. This _should_ give you something like `/path/to/elemental/lib/python2.7/site-packages`. Note that you don't want to install things into the actual /path/to/python/` installation directory because then you wouldn't be able to uninstall them. That is what `spack activate` is for.

**Comment by citibeth** [comment]
> How do I set the PYTHON_SITE_PACKAGES variable for elemental. Basically, I need to know where spack is installing python packages so that I can give that to the build environment. I assume that this info is available through some spack environment variable, but I don’t know which one.

This is really stuff the build should be doing, not Spack.  I use the following in my CMake build that installs Python stuff (this is the standard):
```
install(TARGETS mymodule_so DESTINATION lib/python${PYTHON_VERSION}/site-packages)
```
Notice that you need ${PYTHON_VERSION} --- which should be of the form <major>.<minor>.  To be robust, that must be the version string that *comes from Python*, not whatever Spack thinks it is.  it is obtained in CMake via:
```
    execute_process( COMMAND ${PYTHON_EXECUTABLE} -c "import sysconfig; print (sysconfig.get_python_version())"
                     OUTPUT_VARIABLE PYTHON_VERSION
                     RESULT_VARIABLE PYTHON_VERSION_NOT_FOUND
OUTPUT_STRIP_TRAILING_WHITESPACE)
```
I really suggest you upgrade the Elemental CMake build to query Python, and get it merged upstream.  If you don't, then you will have to do essentially the same thing (plus re-implement `FindPythonInterp.cmake`) in your Spack package --- and it won't be as useful in the end. While you are waiting for them to merge, you can use a temporary Spack-only release in your Spack package; see https://spack.readthedocs.io/en/latest/workflows.html#upstream-bug-fixes

Here is the `FindPython.cmake` I used in my Python3-based projects.  Note that this also tells you all sorts of things *inside* the Python distro; in case you need to include header files, etc.  That would almost certainly also be useful for Elemental, they need to know where `Python.h` is.
https://github.com/citibeth/icebin/blob/develop/cmake/FindPython.cmake

Here is where I use that:
https://github.com/citibeth/icebin/blob/develop/pylib/CMakeLists.txt

Here is bakground on the `sysconfig` module used.  Note that is will not work with Python 2.6.
https://docs.python.org/2/library/sysconfig.html#module-sysconfig

BTW... you should have the following in your Spack package:
```
extends('python')
depends_on('python@:2.8')
```

**Comment by bvanessen** [comment]
Great.  Thanks.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by bvanessen** [comment]
Thank you for the suggestion.  At this point I am really more interested in getting a tool that depends on Elemental into spack, rather than optimizing the build environment for Elemental, but I will save this for a future opportunity.

Adam had a suggestion about how to get the location that I wanted using the site_packages_dir field.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by citibeth** [comment]
@bvanessen The `site_packages_dir` variable is currently undocumented, and I didn't know about it when I wrote the above instructions.  That said... it looks like a reasonable way to go, pending questions about it (see #3008).  Whatever its shortcomings may or may not be, you aren't responsible for fixing them.

Also, I'm wondering how well my build would work in a non-Spack environment...

**Comment by bvanessen** [comment]
Thank you for all of the help.



Brian Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) ‭925-290-5470‬
________________________________

**Comment by davydden** [comment]
@bvanessen you need to fix indentation:
```
var/spack/repos/builtin/packages/elemental/package.py:29: [E501] line too long (104 > 79 characters)
var/spack/repos/builtin/packages/elemental/package.py:110: [E501] line too long (83 > 79 characters)
var/spack/repos/builtin/packages/elemental/package.py:114: [E126] continuation line over-indented for hanging indent
var/spack/repos/builtin/packages/elemental/package.py:118: [E126] continuation line over-indented for hanging indent
var/spack/repos/builtin/packages/elemental/package.py:120: [E501] line too long (85 > 79 characters)
var/spack/repos/builtin/packages/elemental/package.py:127: [E126] continuation line over-indented for hanging indent
var/spack/repos/builtin/packages/elemental/package.py:131: [E501] line too long (88 > 79 characters)
```

**Comment by tgamblin** [comment]
Thanks @bvanessen!

**Comment by tgamblin** [comment]
@poulson: FYI, @elemental will be in Spack pretty soon here.

**Comment by tgamblin** [comment]
@bvanessen: I took a stab at simplifying the CMake args a bit.  Note that you don't have to go full ternary operator on the booleans for CMake -- CMake will accept `True` and `False`, which are the string form of `bool`s in Python.  So you can do this:

    '-DEL_DISABLE_PARMETIS:BOOL=%s' % ('~parmetis' in spec),

Instead of:

    '-DEL_DISABLE_PARMETIS:BOOL={0}'.format(('ON' if '+disable_parmetis' in self.spec else 'OFF')),

**Comment by tgamblin** [comment]
@bvanessen: I also cleaned up the flake8 stuff some.  I think this should be ready to go if the tests pass.  Do you want to test with my changes first?

**Comment by bvanessen** [comment]
I will check tonight.



Brian Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) ‭925-290-5470‬
________________________________

**Comment by tgamblin** [comment]
@bvanessen: FYI, `variant()` lines and the other directives are exempt from flake8 checks, but up to you how you want to format that.

**Comment by bvanessen** [comment]
Oh, okay.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470
